core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -0,0 +1,314 @@
+#![allow(clippy::module_name_repetitions)]
+
+//! This module contains permissions related Iroha functionality.
+
+use std::{fmt::Debug, marker::PhantomData, sync::Arc};
+
+pub use checks::*;
+pub use combinators::ValidatorApplyOr as _;
+use error::*;
+pub use has_token::*;
+use iroha_data_model::prelude::*;
+use iroha_macro::FromVariant;
+pub use is_allowed::*;
+use serde::{Deserialize, Serialize};
+
+#[cfg(test)]
+use crate::wsv::MockWorld;
+use crate::wsv::{World, WorldStateView, WorldTrait};
+
+pub mod builder;
+mod checks;
+pub mod combinators;
+mod has_token;
+mod is_allowed;
+pub mod roles;
+
+/// Result type for permission validators
+pub type Result<T> = std::result::Result<T, DenialReason>;
+
+/// Operation for which the permission should be checked.
+pub trait NeedsPermission: Debug {}
+
+impl NeedsPermission for Instruction {}
+
+impl NeedsPermission for QueryBox {}
+
+// Expression might contain a query, therefore needs to be checked.
+impl NeedsPermission for Expression {}
+
+/// Type of object validator can check
+#[derive(Debug, Copy, Clone, PartialEq, Eq, derive_more::Display)]
+pub enum ValidatorType {
+    /// [`Instruction`] variant
+    Instruction,
+    /// [`QueryBox`] variant
+    Query,
+    /// [`Expression`] variant
+    Expression,
+}
+
+impl ValidatorType {
+    /// Checks if `self` equals to `another`
+    ///
+    /// # Errors
+    /// If `self` doesn't equal to `another`
+    pub fn check_equal(
+        self,
+        another: ValidatorType,
+    ) -> std::result::Result<(), ValidatorTypeMismatch> {
+        if self != another {
+            return Err(ValidatorTypeMismatch {
+                expected: another,
+                actual: self,
+            });
+        }
+
+        Ok(())
+    }
+}
+
+pub mod error {
+    //! Contains errors structures
+
+    use super::ValidatorType;
+    use crate::smartcontracts::Mismatch;
+
+    /// Wrong validator expectation error
+    ///
+    /// I.e. used when user tries to validate [`QueryBox`](super::QueryBox) with
+    /// [`IsAllowedBoxed`](super::IsAllowedBoxed) containing
+    /// [`IsAllowedBoxed::Instruction`](super::IsAllowedBoxed::Instruction) variant
+    pub type ValidatorTypeMismatch = Mismatch<ValidatorType>;
+
+    /// Reason for prohibiting the execution of the particular instruction.
+    #[derive(Debug, Clone, thiserror::Error)]
+    #[allow(variant_size_differences)]
+    pub enum DenialReason {
+        /// [`ValidatorTypeMismatch`] variant
+        #[error("Wrong validator type: {0}")]
+        ValidatorTypeMismatch(#[from] ValidatorTypeMismatch),
+        /// Variant for custom error
+        #[error("{0}")]
+        Custom(String),
+        /// Variant used when at least one [`Validator`](super::IsAllowed) should be provided
+        #[error("No validators provided")]
+        NoValidatorsProvided,
+    }
+
+    impl From<String> for DenialReason {
```

Comment:
how about implementing `FromStr` instead?

Reply:
it might conflict with `thiserror`?

Reply:
I don't think it may help. It is often used with runtime formatted `String`.
So with `FromStr` it will be dereferenced as `&str` and then I have to clone it on heap with `to_owned()`. So it will be double memory allocation? I think it's not good. Better to have one allocation and one move as it is now

Reply:
ok, as you wish. I don't think performance is crucial here for the construction of error variant and `FromStr` seemed more idiomatic to me

Reply:
Hm.... Ok, let's do more ideomatic

Reply:
I think, let's keep both

Reply:
Because `FromStr` returns `Result` which is not very useful here